pkg/disttask/framework/scheduler/collector.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/disttask/framework/proto"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+)
+
+type collector struct {
+	ctx context.Context
+	Param
+
+	subtasks        *prometheus.Desc
+	subtaskDuration *prometheus.Desc
+}
+
+func newCollector(ctx context.Context, param Param) *collector {
+	return &collector{
+		ctx:   ctx,
+		Param: param,
+
+		subtasks: prometheus.NewDesc(
+			"tidb_disttask_subtasks",
+			"Number of subtasks.",
+			[]string{"task_type", "task_id", "status", "exec_id"}, nil,
+		),
+		subtaskDuration: prometheus.NewDesc(
+			"tidb_disttask_subtask_duration",
+			"Duration of subtasks in different states.",
+			[]string{"task_type", "task_id", "status", "subtask_id", "exec_id"}, nil,
+		),
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.subtasks
+	ch <- c.subtaskDuration
+}
+
+// Collect implements the prometheus.Collector interface.
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	subtasks, err := c.taskMgr.GetAllSubtasks(c.ctx)
```

Comment:
this part should move out, if we block on sql for more than scape interval, all metric will be missed